File: js/src/sandbox_fs_utils.ts
import { posix } from "node:path";

import { ClientError, Status } from "nice-grpc";

import {
  InvalidError,
  NotFoundError,
  SandboxFilesystemDirectoryNotEmptyError,
  SandboxFilesystemError,
  SandboxFilesystemFileTooLargeError,
  SandboxFilesystemIsADirectoryError,
  SandboxFilesystemNotADirectoryError,
  SandboxFilesystemNotFoundError,
  SandboxFilesystemPathAlreadyExistsError,
  SandboxFilesystemPermissionError,
} from "./errors";
import { newLogger } from "./logger";

// Module-level logger for diagnostic debug output on failure paths.
const logger = newLogger();

export const SANDBOX_FS_TOOLS_PATH = "/__modal/.bin/modal-sandbox-fs-tools";

export function validateAbsoluteRemotePath(
  remotePath: string,
  operation: string,
): void {
  if (!posix.isAbsolute(remotePath)) {
    throw new InvalidError(
      `Sandbox.filesystem.${operation}() currently only supports absolute remotePath values`,
    );
  }
}

export interface ErrorPayload {
  error_kind: string;
  message: string;
  detail: string;
}

function stderrToText(stderr: string | Uint8Array): string {
  if (typeof stderr === "string") return stderr.trim();
  return new TextDecoder("utf-8", { fatal: false }).decode(stderr).trim();
}

export function tryParseErrorPayload(
  stderr: string | Uint8Array,
): ErrorPayload | null {
  const text = stderrToText(stderr);
  if (!text) return null;

  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch {
    return null;
  }
  if (typeof payload !== "object" || payload === null) return null;

  const obj = payload as Record<string, unknown>;
  const errorKind = obj.error_kind;
  const message = obj.message;
  if (typeof errorKind !== "string") return null;
  if (typeof message !== "string" || !message.trim()) return null;

  const detailRaw = obj.detail;
  const detail = typeof detailRaw === "string" ? detailRaw : "";

  return { error_kind: errorKind, message, detail };
}

// All make*Command outputs must match the `Command` enum in the
// modal-sandbox-fs-tools Rust crate (crates/modal-sandbox-fs-tools/src/lib.rs).
// Treat changes to this schema like protobuf changes: fields must not be
// removed or renamed, only added with backwards-compatible defaults.

export function raiseListFilesError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools listFiles error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (
      payload.error_kind === "IsFile" ||
      payload.error_kind === "NotDirectory"
    )
      throw new SandboxFilesystemNotADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeListFilesCommand(remotePath: string): string {
  return JSON.stringify({ ListFiles: { path: remotePath } });
}

export function raiseMakeDirectoryError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools makeDirectory error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PathAlreadyExists")
      throw new SandboxFilesystemPathAlreadyExistsError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "NotDirectory")
      throw new SandboxFilesystemNotADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "NotSupported")
      throw new InvalidError(
        `${payload.message}: ${remotePath} - this operation is not supported for CloudBucketMounts`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeMakeDirectoryCommand(
  remotePath: string,
  createParents: boolean,
): string {
  return JSON.stringify({
    MakeDirectory: { path: remotePath, parents: createParents },
  });
}

export function raiseReadFileError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools read error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "IsDirectory")
      throw new SandboxFilesystemIsADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "FileTooLarge")
      throw new SandboxFilesystemFileTooLargeError(
        `${payload.message}: ${remotePath}`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeReadFileCommand(remotePath: string): string {
  return JSON.stringify({ ReadFile: { path: remotePath } });
}

export function raiseRemoveError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools remove error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "DirectoryNotEmpty")
      throw new SandboxFilesystemDirectoryNotEmptyError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "NotSupported")
      throw new InvalidError(
        `${payload.message}: ${remotePath} - this operation is not supported for CloudBucketMounts`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeRemoveCommand(
  remotePath: string,
  recursive: boolean,
): string {
  return JSON.stringify({ Remove: { path: remotePath, recursive } });
}

export function raiseStatError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools stat error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "NotDirectory")
      throw new SandboxFilesystemNotADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeStatCommand(remotePath: string): string {
  return JSON.stringify({ Stat: { path: remotePath } });
}

export function raiseWatchError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools watch error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (payload.error_kind === "NotFound")
      throw new SandboxFilesystemNotFoundError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "NotSupported")
      throw new InvalidError(
        `${payload.message}: ${remotePath} - this operation is not supported for CloudBucketMounts`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeWatchCommand(
  remotePath: string,
  opts: {
    recursive: boolean;
    filter: string[] | null;
    timeoutMs: number | null;
  },
): string {
  return JSON.stringify({
    Watch: {
      path: remotePath,
      recursive: opts.recursive,
      filter: opts.filter,
      // The fs-tools wire protocol takes whole seconds; convert from the SDK's
      // millisecond convention, truncating any sub-second remainder.
      timeout_secs:
        opts.timeoutMs !== null ? Math.trunc(opts.timeoutMs / 1000) : null,
    },
  });
}

export function raiseWriteFileError(
  returnCode: number,
  stderr: string | Uint8Array,
  remotePath: string,
): never {
  const payload = tryParseErrorPayload(stderr);
  if (payload) {
    logger.debug(
      `sandbox-fs-tools write error: path=${remotePath}, ` +
        `error_kind=${payload.error_kind}, message=${payload.message}, detail=${payload.detail}`,
    );
    if (
      payload.error_kind === "NotDirectory" ||
      payload.error_kind === "AlreadyExists"
    )
      throw new SandboxFilesystemNotADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "IsDirectory")
      throw new SandboxFilesystemIsADirectoryError(
        `${payload.message}: ${remotePath}`,
      );
    if (payload.error_kind === "PermissionDenied")
      throw new SandboxFilesystemPermissionError(
        `${payload.message}: ${remotePath}`,
      );
    throw new SandboxFilesystemError(payload.message);
  }
  const text = stderrToText(stderr);
  if (text) logger.debug(`Unstructured modal-sandbox-fs-tools stderr: ${text}`);
  throw new SandboxFilesystemError(
    `Operation on '${remotePath}' failed with exit code ${returnCode}`,
  );
}

export function makeWriteFileCommand(remotePath: string): string {
  return JSON.stringify({ WriteFile: { path: remotePath } });
}

const SUPPORT_ERROR_CODE_RE = /Error code:\s*([A-Z0-9]{8})/;

export function extractSupportErrorCode(err: unknown): string | null {
  const message = err instanceof Error ? err.message : String(err);
  const match = SUPPORT_ERROR_CODE_RE.exec(message);
  return match ? match[1] : null;
}

function describeError(err: unknown): string {
  if (err instanceof Error) return `${err.name}: ${err.message}`;
  return String(err);
}

/** NotFound and Service Errors */
function isSandboxUnavailableError(err: unknown): boolean {
  if (err instanceof NotFoundError) return true;
  if (err instanceof ClientError) {
    return (
      err.code === Status.NOT_FOUND ||
      err.code === Status.CANCELLED ||
      err.code === Status.UNKNOWN ||
      err.code === Status.DEADLINE_EXCEEDED ||
      err.code === Status.UNAVAILABLE
    );
  }
  return false;
}

function translateExecSandboxUnavailableError(
  operation: string,
  remotePath: string,
  err: unknown,
): Error {
  logger.debug(
    `Sandbox filesystem control-plane error for operation=${operation}, path=${remotePath}: ${describeError(err)}`,
  );
  return new NotFoundError(
    "The Sandbox is unavailable. This Sandbox may have already shut down.",
  );
}

/**
 * Translate an unexpected exec-level error into a generic user-facing error.
 *
 * This is less than ideal — it discards the original exception and returns a
 * generic message — but it's necessary to avoid surfacing "call to exec()
 * failed"-style messages to users who shouldn't be aware that the filesystem
 * API is implemented in terms of Sandbox.exec().
 */
export function translateExecUnexpectedError(
  operation: string,
  remotePath: string,
  err: unknown,
): Error {
  const code = extractSupportErrorCode(err);
  logger.debug(
    `Unexpected sandbox filesystem exec error for operation=${operation}, path=${remotePath}: ${describeError(err)}`,
  );
  const supportSuffix = code
    ? `please contact support@modal.com (Error code: ${code})`
    : "please contact support@modal.com";
  return new SandboxFilesystemError(
    `An unexpected error occurred, ${supportSuffix}`,
  );
}

/** Translate exec-level exceptions into user-facing errors. */
export async function translateExecErrors<T>(
  operation: string,
  remotePath: string,
  fn: () => Promise<T>,
): Promise<T> {
  try {
    return await fn();
  } catch (err) {
    if (isSandboxUnavailableError(err)) {
      throw translateExecSandboxUnavailableError(operation, remotePath, err);
    }
    if (err instanceof ClientError) {
      throw translateExecUnexpectedError(operation, remotePath, err);
    }
    throw err;
  }
}

File: go/sandbox_fs_utils.go
package modal

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"regexp"
	"strings"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

var supportErrorCodeRe = regexp.MustCompile(`Error code:\s*([A-Z0-9]{8})`)

type errorPayload struct {
	ErrorKind string `json:"error_kind"`
	Message   string `json:"message"`
	Detail    string `json:"detail"`
}

func tryParseErrorPayload(stderr []byte) *errorPayload {
	text := strings.TrimSpace(string(stderr))
	if text == "" {
		return nil
	}
	var p errorPayload
	if err := json.Unmarshal([]byte(text), &p); err != nil {
		return nil
	}
	if p.ErrorKind == "" || strings.TrimSpace(p.Message) == "" {
		return nil
	}
	return &p
}

func isSandboxUnavailable(err error) bool {
	if _, ok := status.FromError(err); !ok {
		// status.Code returns codes.Unknown for any non-gRPC error, so guard
		// against treating local OS errors (e.g. a failed local file write) as
		// "sandbox unavailable".
		return false
	}
	switch status.Code(err) {
	case codes.NotFound, codes.Canceled, codes.Unknown, codes.DeadlineExceeded, codes.Unavailable:
		return true
	}
	return false
}

func isBinaryExitedEarly(err error) bool {
	c := status.Code(err)
	return c == codes.FailedPrecondition || c == codes.Aborted
}

func extractSupportErrorCode(err error) string {
	if m := supportErrorCodeRe.FindStringSubmatch(err.Error()); len(m) > 1 {
		return m[1]
	}
	return ""
}

func translateExecError(ctx context.Context, log *slog.Logger, operation, remotePath string, err error) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}
	if isSandboxUnavailable(err) {
		log.DebugContext(ctx, "Sandbox filesystem control-plane error",
			"operation", operation,
			"path", remotePath,
			"error", err,
		)
		return NotFoundError{Exception: "The Sandbox is unavailable. This Sandbox may have already shut down."}
	}
	log.DebugContext(ctx, "Unexpected sandbox filesystem exec error",
		"operation", operation,
		"path", remotePath,
		"error", err,
	)
	if code := extractSupportErrorCode(err); code != "" {
		return SandboxFilesystemError{Exception: fmt.Sprintf("An unexpected error occurred, please contact support@modal.com (Error code: %s)", code)}
	}
	return SandboxFilesystemError{Exception: "An unexpected error occurred, please contact support@modal.com"}
}

func validateAbsoluteRemotePath(remotePath, operation string) error {
	if !strings.HasPrefix(remotePath, "/") {
		return InvalidError{Exception: fmt.Sprintf("Sandbox.filesystem.%s() currently only supports absolute remote_path values", operation)}
	}
	return nil
}

func raiseListFilesError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools list_files error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "IsFile", "NotDirectory":
			return SandboxFilesystemNotADirectoryError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeListFilesCommand(remotePath string) string {
	b, _ := json.Marshal(map[string]any{"ListFiles": map[string]any{"path": remotePath}})
	return string(b)
}

func raiseMakeDirectoryError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools make_directory error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "PathAlreadyExists":
			return SandboxFilesystemPathAlreadyExistsError{Exception: p.Message + ": " + remotePath}
		case "NotDirectory":
			return SandboxFilesystemNotADirectoryError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		case "NotSupported":
			return InvalidError{Exception: p.Message + ": " + remotePath + " - this operation is not supported for CloudBucketMounts"}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeMakeDirectoryCommand(remotePath string, createParents bool) string {
	b, _ := json.Marshal(map[string]any{"MakeDirectory": map[string]any{"path": remotePath, "parents": createParents}})
	return string(b)
}

func raiseReadFileError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools read error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "IsDirectory":
			return SandboxFilesystemIsADirectoryError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		case "FileTooLarge":
			return SandboxFilesystemFileTooLargeError{Exception: p.Message + ": " + remotePath}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeReadFileCommand(remotePath string) string {
	b, _ := json.Marshal(map[string]any{"ReadFile": map[string]any{"path": remotePath}})
	return string(b)
}

func raiseRemoveError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools remove error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "DirectoryNotEmpty":
			return SandboxFilesystemDirectoryNotEmptyError{Exception: p.Message + ": " + remotePath}
		case "NotSupported":
			return InvalidError{Exception: p.Message + ": " + remotePath + " - this operation is not supported for CloudBucketMounts"}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeRemoveCommand(remotePath string, recursive bool) string {
	b, _ := json.Marshal(map[string]any{"Remove": map[string]any{"path": remotePath, "recursive": recursive}})
	return string(b)
}

func raiseStatError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools stat error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "NotDirectory":
			return SandboxFilesystemNotADirectoryError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeStatCommand(remotePath string) string {
	b, _ := json.Marshal(map[string]any{"Stat": map[string]any{"path": remotePath}})
	return string(b)
}

func raiseWatchError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools watch error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotFound":
			return SandboxFilesystemNotFoundError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		case "NotSupported":
			return InvalidError{Exception: p.Message + ": " + remotePath + " - this operation is not supported for CloudBucketMounts"}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeWatchCommand(remotePath string, recursive bool, filter []string, timeoutSecs *int) string {
	var filterVal any
	if filter != nil {
		filterVal = filter
	}
	var timeoutVal any
	if timeoutSecs != nil {
		timeoutVal = *timeoutSecs
	}
	b, _ := json.Marshal(map[string]any{"Watch": map[string]any{"path": remotePath, "recursive": recursive, "filter": filterVal, "timeout_secs": timeoutVal}})
	return string(b)
}

func raiseWriteFileError(ctx context.Context, log *slog.Logger, returnCode int, stderr []byte, remotePath string) error {
	if p := tryParseErrorPayload(stderr); p != nil {
		log.DebugContext(ctx, "sandbox-fs-tools write error",
			"path", remotePath,
			"error_kind", p.ErrorKind,
			"message", p.Message,
			"detail", p.Detail,
		)
		switch p.ErrorKind {
		case "NotDirectory", "AlreadyExists":
			return SandboxFilesystemNotADirectoryError{Exception: p.Message + ": " + remotePath}
		case "IsDirectory":
			return SandboxFilesystemIsADirectoryError{Exception: p.Message + ": " + remotePath}
		case "PermissionDenied":
			return SandboxFilesystemPermissionError{Exception: p.Message + ": " + remotePath}
		default:
			return SandboxFilesystemError{Exception: p.Message}
		}
	}
	if text := strings.TrimSpace(string(stderr)); text != "" {
		log.DebugContext(ctx, "Unstructured modal-sandbox-fs-tools stderr", "stderr", text)
	}
	return SandboxFilesystemError{Exception: fmt.Sprintf("Operation on '%s' failed with exit code %d", remotePath, returnCode)}
}

func makeWriteFileCommand(remotePath string) string {
	b, _ := json.Marshal(map[string]any{"WriteFile": map[string]any{"path": remotePath}})
	return string(b)
}
